### PR TITLE
Allow affix settings to specify dependencies

### PR DESCRIPTION
**Pull request by s1monw**
We use affix settings to group settings / values under a certain namespace.
In some cases like login information for instance a setting is only valid if
one or more other settings are present. For instance `x.test.user` is only valid
if there is an `x.test.passwd` present and vice versa. This change allows to specify
such a dependency to prevent settings updates that leave settings in an inconsistent
state.

### CODE REVIEW

**Comment by javanna** [comment]
@s1monw I have been trying this out in #27182. Do I understand correctly that a setting which is dependent on another setting must always be provided in the same request as its dependency? That makes sense for user and password, but in CCS one may want to change the `skip_if_disconnected` setting without touching the seeds, in which case an error is thrown as seeds haven't been specified. Thoughts?

You can reproduce with the following unit test:

```
AbstractScopedSettings service = new ClusterSettings(Settings.builder().put("search.remote.foo.seeds", seed).build(),
                        new HashSet<>(Arrays.asList(RemoteClusterAware.REMOTE_CLUSTERS_SEEDS,
                                RemoteClusterService.REMOTE_CLUSTER_SKIP_IF_DISCONNECTED)));
                service2.validate(Settings.builder().put("search.remote.foo.skip_if_disconnected", randomBoolean()));
```

**Comment by s1monw** [comment]
@javanna I applied your feedback thx

**Comment by s1monw** [comment]
@elasticmachine test this please

**Comment by s1monw** [comment]
@javanna can you take another look?

File: core/src/main/java/org/elasticsearch/action/admin/cluster/settings/SettingsUpdater.java [hunk 1]
```diff
@@ -54,15 +54,23 @@ synchronized ClusterState updateSettings(final ClusterState currentState, Settin
...
+
         Settings.Builder persistentSettings = Settings.builder();
         persistentSettings.put(currentState.metaData().persistentSettings());
         changed |= clusterSettings.updateDynamicSettings(persistentToApply, persistentSettings, persistentUpdates, "persistent");
 
         final ClusterState clusterState;
         if (changed) {
+            Settings transientFinalSettings = transientSettings.build();
+            Settings persistentFinalSettings = persistentSettings.build();
+            // both transient and persistent settings must be consistent by itself we can't allow dependencies to be
```
[the remark this answers is not in the document]
Comment (javanna) [reply core/src/main/java/org/elasticsearch/action/admin/cluster/settings/SettingsUpdater.java]: "we can't allow dependencies between the two"?

**Review by javanna — APPROVED** [review]
I tested it out and it works well with the changes from #27182 . I left a few minor comments and a couple of questions. The only minor thing that I wish we could improve, but I am not sure we can, is the error message when you set to null a setting that another one left behind depends on. It may not be clear to users what they have to do to fix that (e.g. set to null the other setting too).

File: core/src/main/java/org/elasticsearch/action/admin/cluster/settings/SettingsUpdater.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (s1monw) [reply core/src/main/java/org/elasticsearch/action/admin/cluster/settings/SettingsUpdater.java]: what if you have setting A to be transient and then it disappears on a full cluster restart?

File: core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataUpdateSettingsService.java [hunk 2]
```diff
@@ -240,7 +241,9 @@ public ClusterState execute(ClusterState currentState) {
                             if (preserveExisting) {
                                 indexSettings.put(indexMetaData.getSettings());
                             }
-                            metaDataBuilder.put(IndexMetaData.builder(indexMetaData).settings(indexSettings));
+                            Settings finalSettings = indexSettings.build();
+                            indexScopedSettings.validate(finalSettings.filter(k -> indexScopedSettings.isPrivateSetting(k) == false), true);
```
[the remark this answers is not in the document]
Comment (s1monw) [reply core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataUpdateSettingsService.java]: yeah

File: core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataUpdateSettingsService.java [hunk 3]
```diff
@@ -240,7 +241,9 @@ public ClusterState execute(ClusterState currentState) {
...
+                            metaDataBuilder.put(IndexMetaData.builder(indexMetaData).settings(finalSettings));
```
[the remark this answers is not in the document]
Comment (s1monw) [reply core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataUpdateSettingsService.java]: one happens before we pass on to a clusterstate update task we should validate as soon as we can. but we have to do it here again

File: plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureStorageSettings.java [hunk 4]
```diff
@@ -41,41 +41,42 @@
...
-        (key) -> Setting.timeSetting(key, TimeValue.timeValueMinutes(-1), Property.NodeScope));
+    public static final AffixSetting<TimeValue> TIMEOUT_SETTING = Setting.affixKeySetting("azure.client.", "timeout",
+        (key) -> Setting.timeSetting(key, TimeValue.timeValueMinutes(-1), Property.NodeScope), ACCOUNT_SETTING, KEY_SETTING);
 
     /** The type of the proxy to connect to azure through. Can be direct (no proxy, default), http or socks */
-    public static final AffixSetting<Proxy.Type> PROXY_TYPE_SETTING = Setting.affixKeySetting(PREFIX, "proxy.type",
-        (key) -> new Setting<>(key, "direct", s -> Proxy.Type.valueOf(s.toUpperCase(Locale.ROOT)), Property.NodeScope));
+    public static final AffixSetting<Proxy.Type> PROXY_TYPE_SETTING = Setting.affixKeySetting("azure.client.", "proxy.type",
+        (key) -> new Setting<>(key, "direct", s -> Proxy.Type.valueOf(s.toUpperCase(Locale.ROOT)), Property.NodeScope)
+        , ACCOUNT_SETTING, KEY_SETTING);
```
[the remark this answers is not in the document]
Comment (s1monw) [reply plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureStorageSettings.java]: yeah I am not sure about that

File: core/src/main/java/org/elasticsearch/action/admin/cluster/settings/SettingsUpdater.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (javanna) [reply core/src/main/java/org/elasticsearch/action/admin/cluster/settings/SettingsUpdater.java]: sure I agree, I was just suggesting to rephrase the comment :)

File: core/src/main/java/org/elasticsearch/action/admin/indices/template/put/TransportPutIndexTemplateAction.java [hunk 5]
```diff
@@ -77,7 +77,7 @@ protected void masterOperation(final PutIndexTemplateRequest request, final Clus
         }
         final Settings.Builder templateSettingsBuilder = Settings.builder();
         templateSettingsBuilder.put(request.settings()).normalizePrefix(IndexMetaData.INDEX_SETTING_PREFIX);
-        indexScopedSettings.validate(templateSettingsBuilder);
+        indexScopedSettings.validate(templateSettingsBuilder.build(), true); // templates must be consistent with reg. to dependencies
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/action/admin/indices/template/put/TransportPutIndexTemplateAction.java]: Can you expand `reg.`? It took me a few seconds to figure out what it meant (`regards`) and I do not want to have to do it every time I read this comment?

**Review by jasontedor — APPROVED** [review]
This looks good to me, I left one minor request.

File: plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureStorageSettings.java [hunk 6]
```diff
@@ -41,41 +41,41 @@
 import java.util.Set;
 
 public final class AzureStorageSettings {
-    // prefix for azure client settings
-    private static final String PREFIX = "azure.client.";
```
Comment (rjernst) [inline plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureStorageSettings.java]: Why remove the internal constant? Seems like this is just asking for a typo in the future in one setting to cause a storm of problems :/

Comment (s1monw) [reply plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureStorageSettings.java]: fair enough. I can add it back.

**Pull request merged**